components/uArchARM/MapTable.hpp

```diff
@@ -225,28 +133,24 @@ struct PhysicalMap {
   void free(pRegister aRegisterName) {
     FLEXUS_PROFILE();
     theFreeList.push_back(aRegisterName);
-    std::list<pRegister>::iterator iter, end;
     int32_t arch_name = theReverseMappings[aRegisterName];
+    std::vector<std::list<pRegister>>::iterator iter = theAssignedRegisters.begin() + arch_name;
     DBG_Assert(arch_name >= 0 && arch_name < static_cast<int>(theAssignedRegisters.size()));
     DBG_Assert(theMappings[arch_name] != aRegisterName);
-    DBG_Assert(theAssignedRegisters[arch_name].size());
+    DBG_Assert(iter->size());
```

Comment:
Do you think it's better to only include these asserts at higher debug levels? (e.g., Trace and greater?)

As is, it seems like they would take up significant processing time

Reply:
I don't understand.
What do you mean for trace and greater? These are asserts and not just debug messages.
I would like Flexus to crash if these fail at any point.

I guess that these asserts should actually provide better performance.
The original code right after this comment was going through the whole list to find the required mappings.
But now it can directly find the required mapping from the end of the list, which is better.

Reply:
Two things:
1) You can use DBG_AssertSev( ) to only enable asserts at a certain debug level or above. I do this often for correctness checks which are on the critical path, and then compile them out below a certain level.

2) The code that we replaced seems to perform a completely different functionality than before. They are searching through the whole mapping list to find the right one, and we are now assuming it's always at the front or back. Is there some change we made (e.g., removing the `rotate()` call) that now ensures we are always at front or back?

Reply:
1. Ah, alright. I didn't know this. I would still keep these asserts there always, and I do not think they are a big performance overhead as well.

2. The overall logic for mapping is the one we are using now. Oldest mappings are always at the front of this queue, and the newest ones are at the back. We are guaranteed to always erase mappings either from the front (when retiring instructions and making the mappings permanent) or the back (when squashing instructions).
So, even though the previous code did the same thing in a more generic way, then current implementation is more efficient.

Reply:
1) If you want to keep them, OK.

2) The logic is intuitive, but it's very weird that this was not so straightforwards before. Was it possible in Simflex that a mapping for a register (e.g., %l4) would exist in a different window, and therefore a linear search was necessary? I guess if we move all calls to `rotate()` in map/free/restore, we should not need the search.